### PR TITLE
Use numerically stable one-pass algorithm for log-sum-exp

### PR DESCRIPTION
**Pull request by devmotion**
A while ago I wrote a Julia implementation of a numerically stable one-pass algorithm for log-sum-exp (original PR: https://github.com/JuliaStats/StatsFuns.jl/pull/97, later extracted to https://github.com/JuliaStats/LogExpFunctions.jl/blob/master/src/logsumexp.jl). This one-pass algorithm is faster (at least for the inputs I benchmarked) and also fixes some underflow issues of the standard implementation of log-sum-exp (that are present in Birch as well). Hence I thought it might be useful for Birch as well and switched `log_sum_exp` and `resample_reduce` in the standard library to the one-pass algorithm.

I checked the implementation with the following Birch code: https://gist.github.com/devmotion/a6c3561f6c593160744147e7c5165f62

The output indicates that the PR fixes the underflow issue and improves performance:
```shell
❯ birch log_sum_exp_underflow -onepass false
f: log_sum_exp
f([1e-20, log(1e-20)]) = 1.00000000000000e-20 (correct: ~1.999999999999999999985e-20)
❯ birch log_sum_exp_underflow -onepass true
f: log_sum_exp_onepass
f([1e-20, log(1e-20)]) = 2.00000000000000e-20 (correct: ~1.999999999999999999985e-20)
❯ birch resample_reduce_underflow -onepass false
f: resample_reduce
f([1e-20, log(1e-20)]) = (1.0, 1.00000000000000e-20) (correct: (_, ~1.999999999999999999985e-20))
❯ birch resample_reduce_underflow -onepass true
f: resample_reduce_onepass
f([1e-20, log(1e-20)]) = (1.0, 2.00000000000000e-20) (correct: (_, ~1.999999999999999999985e-20))
❯ birch log_sum_exp_timings
current: 7.37634658271839e+00 (result), 1.28010000000000e-05 (time)
onepass: 7.37634658271839e+00 (result), 6.99000000000000e-06 (time)
❯ birch log_sum_exp_timings
current: 7.40357961126728e+00 (result), 1.23700000000000e-05 (time)
onepass: 7.40357961126728e+00 (result), 6.95700000000000e-06 (time)
❯ birch resample_reduce_timings                
current: (3.40619836613369e+02, 7.41604189731166e+00) (result), 1.69120000000000e-05 (time)
onepass: (3.40619836613369e+02, 7.41604189731166e+00) (result), 7.15000000000000e-06 (time)
❯ birch resample_reduce_timings
current: (2.37522674833629e+02, 7.43909556079362e+00) (result), 1.29710000000000e-05 (time)
onepass: (2.37522674833630e+02, 7.43909556079362e+00) (result), 8.29000000000000e-06 (time)
```

I guess it might be good to add the underflow example and possibly some other tests. Where should they be added?

### CODE REVIEW

**Comment by lawmurray** [comment]
Looks nice @devmotion, thanks for contributing this. I'm rebasing to the numeric branch, which has been long running, as I intend to merge it back to master soon. That will also run the full test suite, specifically so we can confirm that the log-normalizing constants for the various examples are consistent.

For tests, you can add them to `libraries/StandardTest/src/basic` on master, or `tests/Test/src/basic` on numeric. They should be consistent with other tests there, as in exit with a nonzero code on failure, in which case they can output some diagnostics, but otherwise stay silent.

Finally, is there a reference to explain the numerical stability aspect of this to a future reader? The [blog post](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html) linked in the code comments has a nice explanation of the streaming part, but not the numerical stability part, which I think you may have added.

**Comment by lawmurray** [comment]
Normalizing constant tests are failing, e.g. `test_z_beta` in https://app.circleci.com/pipelines/github/lawmurray/Birch/1185/workflows/3df77928-2ab6-4577-837c-2cf02b77767f/jobs/18530

These tests are new on branch numeric, and don't exist on master, so this was not caught until now. They're all those beginning with `test_z`. These tests use importance sampling to estimate the normalizing constant of a probability distribution, and use `resample_reduce()` internally, so likely something is at fault there (possibly introduced by the rebase and merge, will need to investigate).

**Comment by devmotion** [comment]
> Looks nice @devmotion, thanks for contributing this. I'm rebasing to the numeric branch, which has been long running, as I intend to merge it back to master soon.

Thanks!

> For tests, you can add them to `libraries/StandardTest/src/basic` on master, or `tests/Test/src/basic` on numeric. They should be consistent with other tests there, as in exit with a nonzero code on failure, in which case they can output some diagnostics, but otherwise stay silent.

OK, I'll add tests when I'm back from vacation next week.

> Finally, is there a reference to explain the numerical stability aspect of this to a future reader? The [blog post](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html) linked in the code comments has a nice explanation of the streaming part, but not the numerical stability part, which I think you may have added.

True, I added that in the Julia implementation but it was not part of the blog post. The idea is to only add the `+ 1` for the maximum element in the end but not during the reduction. I'll add some explanation.

**Comment by devmotion** [comment]
> Normalizing constant tests are failing, e.g. test_z_beta

Oh, I'll check what's the problem there and try to fix it. I wonder if it is caused by the NaN values in `z` resulting from samples outside of the support of the beta distribution. Maybe one has to deal with NaN a bit more carefully in the reduction.

**Comment by lawmurray** [comment]
Thanks for the updates on this @devmotion. I made a few minor tweaks for style (spaces around low precedence operators, rewrote a couple of nested `if`s to look a little nicer). I'll be merging `numeric` back to `master` shortly, but please stay based on `numeric` for final changes and I can do another merge to `master` when you're done to pick up the final changes (e.g. some docs and tests mentioned above). Additional tests, by the way, are not essential, as those normalizing constant tests already cover the functions involved here.

**Comment by lawmurray** [comment]
Also, once you merge updates on `numeric` into the PR, we should see the FreeBSD and Codecov tests passing. I merged your FreeBSD fix onto `numeric` and not just `master` as before, and tweaked the Codecov settings to ignore small changes in coverage (I think!).

**Comment by devmotion** [comment]
Sorry for the slow progress here, I was quite busy last week. I added some explanations and tests (revealed an incorrect result of `resample_reduce` for empty input vectors, it seems).

Indeed, the `if` statements were a bit ugly and unsatisfying. The main idea was to perform as few checks as possible and e.g. check for `inf` only when the next element is larger but not if it is smaller than the current maximum. I guess it's not worth it though (I'm not sure if there are any performance gains at all), and I prefer the simpler and more readable version that you changed it to.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 1]
```diff
@@ -0,0 +1,149 @@
+
+/*
+ * Test log-sum-exp implementations in `log_sum_exp` and
+ * `resample_reduce`.
+ */
+program test_logsumexp() {
```
Comment (lawmurray) [inline tests/Test/src/basic/test_basic_logsumexp.birch]: Should be `test_basic_logsumexp` to be picked up and run by the `grep` in `smoke.sh` and `test.sh`. Or I'd even suggest `test_basic_log_sum_exp` here to be consistent with the name of the function.

Comment (lawmurray) [reply tests/Test/src/basic/test_basic_logsumexp.birch]: Similar throughout, `logsumexp` -> `log_sum_exp`.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+  // Compare with common two-pass algorithm
+  let y <- log_sum_exp_twopass(w);
+  let w2 <- transform(w, \(x:Real) -> { return 2*w; });
+  let ess <- y*y/log_sum_exp_twopass(w2);
+  if !check_ess_logsumexp(w, ess, y) {
+    exit(1);
+  }
+  
+  // Check overflow
```
Comment (lawmurray) [inline tests/Test/src/basic/test_basic_logsumexp.birch:20]: By way of style, use `/* ... */` for whole line comments and no capitals, so `/* check overflow */` here.

Comment (lawmurray) [reply tests/Test/src/basic/test_basic_logsumexp.birch:20]: A few more examples below.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+  }
+
+  // Check empty input
+  x:Real[0];
+  if !check_ess_logsumexp(x, 0.0, -inf) {
+    exit(1);
+  }
+
+  // Special cases involving -inf, inf, and nan.
+  let cases <- [
```
Comment (lawmurray) [inline tests/Test/src/basic/test_basic_logsumexp.birch]: Nice! Thanks for including all these edge cases.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+ *
+ * @param x1 First scalar.
+ * @param x2 Second scalar.
+ * @param reltol Relative tolerace.
+ *
+ * @return Are the two scalars approximately equal?
+ */
+function approx_equal(x1:Real, x2:Real, reltol:Real) -> Boolean {
+  // Handle special cases such as `inf`, `-inf` etc. not covered below.
+  if x1==x2 {
```
Comment (lawmurray) [inline tests/Test/src/basic/test_basic_logsumexp.birch:143]: Use spaces around low-precedence operators, meaning anything but `*` and `/`. Thus `x1 == x2` here.

File: libraries/Standard/src/primitive/resample.birch [hunk 5]
```diff
@@ -35,20 +35,31 @@ function nan_max(w:Real[_]) -> Real {
...
+    /*
+     * Running sum of non-maximum weights divided by maximum weight.
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
     wn:Real;
     for n in 1..length(w) {
       wn <- w[n];
```
Comment (lawmurray) [inline libraries/Standard/src/primitive/resample.birch:65]: Use `let wn <- w[n]` here.

File: libraries/Standard/src/primitive/resample.birch [hunk 6]
```diff
@@ -35,20 +35,31 @@ function nan_max(w:Real[_]) -> Real {
...
+    // Running maximum of log weights
+    let mx <- -inf;
+    /*
+     * Running sum of non-maximum weights divided by maximum weight.
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
     wn:Real;
```
Comment (lawmurray) [inline libraries/Standard/src/primitive/resample.birch:63]: Delete, use `let wn <- w[n]` below instead, as `wn` doesn't need to be preserved over iterations of the loop.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+/*
+ * Exponentiate and sum a log weight vector.
+ *
+ * @param w Log weights.
+ *
+ * @return the logarithm of the sum.
+ *
+ * !!! note
```
Comment (lawmurray) [inline tests/Test/src/basic/test_basic_logsumexp.birch:74]: Make it `@note`, but keep the following paragraph indented (it converts to `!!! note` in Markdown anyway, but standardizing on Doxygen-style `@note` toward C++ convergence).

File: libraries/Standard/src/primitive/resample.birch [hunk 8]
```diff
@@ -35,20 +35,31 @@ function nan_max(w:Real[_]) -> Real {
...
  * Exponentiate and sum a log weight vector.
  *
- * The result is computed in a numerically stable way that avoids
- * intermediate over- and underflow, using a single pass over the data.
- *
  * @param w Log weights.
  *
  * @return the logarithm of the sum.
  *
  * !!! note
```
Comment (lawmurray) [inline libraries/Standard/src/primitive/resample.birch:42]: `@note` (explanation on another instance below).

File: libraries/Standard/src/primitive/resample.birch [hunk 9]
```diff
@@ -273,14 +284,29 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+    /*
+     * Running sum of non-maximum weights divided by maximum weight (r),
+     * and their squares (q).
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
+    let q <- 0.0;
     wn:Real;
```
Comment (lawmurray) [inline libraries/Standard/src/primitive/resample.birch:310]: Remove and use `let wn <- ...` below.

File: libraries/Standard/src/primitive/resample.birch [hunk 10]
```diff
@@ -273,14 +284,29 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
     v:Real;
```
Comment (lawmurray) [inline libraries/Standard/src/primitive/resample.birch:311]: Likewise, remove and use `let v <- ...` below.

**Review by lawmurray — CHANGES_REQUESTED** [review]
Thanks @devmotion. Looks solid, only issue is `program` name, see comments. Otherwise, I left a few comments mainly style related, but keen to keep a consistent style throughout the code.

**Comment by lawmurray** [comment]
> Indeed, the `if` statements were a bit ugly and unsatisfying. The main idea was to perform as few checks as possible and e.g. check for `inf` only when the next element is larger but not if it is smaller than the current maximum. I guess it's not worth it though (I'm not sure if there are any performance gains at all), and I prefer the simpler and more readable version that you changed it to.

No worries. Just favoring the neater code, and can expect the compiler to optimize it out here.

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 1 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch]: Ah I had noticed this probably as well and thought I had pushed a commit that fixes the name. It seems I did not :smile:

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 2 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch:20]:
```suggestion
  /* check overflow */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 4 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch:143]:
```suggestion
  if x1 == x2 {
```

File: libraries/Standard/src/primitive/resample.birch (hunk 6 above)
Comment (devmotion) [reply libraries/Standard/src/primitive/resample.birch:63]:
```suggestion
```

File: libraries/Standard/src/primitive/resample.birch (hunk 5 above)
Comment (devmotion) [reply libraries/Standard/src/primitive/resample.birch:65]:
```suggestion
      let wn <- w[n];
```

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 7 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch:74]:
```suggestion
 * @note
```

File: libraries/Standard/src/primitive/resample.birch (hunk 8 above)
Comment (devmotion) [reply libraries/Standard/src/primitive/resample.birch:42]:
```suggestion
 * @note
```

File: libraries/Standard/src/primitive/resample.birch (hunk 9 above)
Comment (devmotion) [reply libraries/Standard/src/primitive/resample.birch:310]:
```suggestion
```

File: libraries/Standard/src/primitive/resample.birch (hunk 10 above)
Comment (devmotion) [reply libraries/Standard/src/primitive/resample.birch:311]:
```suggestion
```

File: libraries/Standard/src/primitive/resample.birch [hunk 11]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+     * In contrast to Nowozin's implementation the maximum itself
+     * is not included in the sum.
+     * This avoids underflow for e.g. w = [1e-20, log(1e-20)].
+     */
+    let r <- 0.0;
+    let q <- 0.0;
+    wn:Real;
+    v:Real;
+    for n in 1..length(w) {
+      wn <- w[n];
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:313]:
```suggestion
      let wn <- w[n];
```

File: libraries/Standard/src/primitive/resample.birch [hunk 12]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+      if wn == inf {
+        return (1.0, inf);
+      } else if wn > mx {
+        v <- exp(mx - wn);
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:317]:
```suggestion
        let v <- exp(mx - wn);
```

File: libraries/Standard/src/primitive/resample.birch [hunk 13]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+        r <- (r + 1.0)*v;
+        q <- (q + 1.0)*v*v;
+        mx <- wn;
+      } else if isfinite(wn) {
+        v <- exp(wn - mx);
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:322]:
```suggestion
        let v <- exp(wn - mx);
```

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 14]
```diff
@@ -0,0 +1,148 @@
+
+/*
+ * Test log-sum-exp implementations in `log_sum_exp` and
+ * `resample_reduce`.
+ */
+program test_basic_logsumexp() {
+  // Generate random weights
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:7]:
```suggestion
  /* generate random weights */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 1 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch]: Oh, actually I think I renamed it to `test_basic_logsumexp` in https://github.com/lawmurray/Birch/pull/18/commits/bc6d13ac0e2fa0fc97f0da926f72f5c6a491e493, so maybe you were looking at an older commit? In any case, I'll rename `logsumexp` to `log_sum_exp`.

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+  w:Real[100];
+  for n in 1..100 {
+    w[n] <- simulate_gaussian(0.0, 1.0);
+  }
+
+  // Compare with common two-pass algorithm
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:13]:
```suggestion
  /* compare with common two-pass algorithm */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+  
+  // Check overflow
+  w <- w + 1000.0;
+  y <- log_sum_exp_twopass(w);
+  ess <- exp(2*y - log_sum_exp_twopass(2*w));
+  if !check_ess_logsumexp(w, ess, y) {
+    exit(1);
+  }
+
+  // Check underflow
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:28]:
```suggestion
  /* check underflow */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+  w <- [1e-20, log(1e-20)];
+  y <- 2e-20;
+  ess <- 1.0;
+  if !check_ess_logsumexp(w, ess, y) {
+    exit(1);
+  }
+
+  // Check empty input
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:36]:
```suggestion
  /* check empty input */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+  x:Real[0];
+  if !check_ess_logsumexp(x, 0.0, -inf) {
+    exit(1);
+  }
+
+  // Special cases involving -inf, inf, and nan.
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:42]:
```suggestion
  /* special cases involving -inf, inf, and nan */
```

File: tests/Test/src/basic/test_basic_logsumexp.birch [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+ * Check output of `log_sum_exp` and `resample_reduce`.
+ *
+ * @param w Log weights.
+ * @param ess_expected Expected ESS estimate.
+ * @param y_expected Expected log sum of weights.
+ *
+ * @return Are the results approximately equal to the expected values?
+ */
+function check_ess_logsumexp(w:Real[_], ess_expected:Real, y_expected:Real) -> Boolean {
+  // Roughly half of the significant digits should be correct.
```
Comment (devmotion) [inline tests/Test/src/basic/test_basic_logsumexp.birch:99]:
```suggestion
  /* roughly half of the significant digits should be correct */
```

File: libraries/Standard/src/primitive/resample.birch [hunk 20]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+        r <- r + v;
+        q <- q + v*v;
+      }
     }
-    return (W*W/W2, log(W) + mx);
+
+    // If all weights are `-inf` or `nan`, the result is the same as for empty arrays
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:328]:
```suggestion
    /* if all weights are `-inf` or `nan`, the result is the same as for empty arrays */
```

File: libraries/Standard/src/primitive/resample.birch [hunk 21]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
+    if mx==-inf {
+      return (0.0, -inf);
+    }
+
+    let log_sum_weights <- mx + log1p(r);
+    // The ESS is estimated as (sum w)^2 / (sum w^2)
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:334]:
```suggestion
    /* the ESS is estimated as (sum w)^2 / (sum w^2) */
```

File: libraries/Standard/src/primitive/resample.birch [hunk 22]
```diff
@@ -41,13 +41,38 @@ function nan_max(w:Real[_]) -> Real {
...
+ * http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html
  */
 function log_sum_exp(w:Real[_]) -> Real {
   if length(w) > 0 {
-    let mx <- nan_max(w);
-    let r <- transform_reduce(w, 0.0, \(x:Real, y:Real) -> { return x + y; },
-        \(x:Real) -> { return nan_exp(x - mx); });
-    return mx + log(r);
+    // Running maximum of log weights
+    let mx <- -inf;
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:55]:
```suggestion
    let mx <- -inf; // running maximum of log weights
```

File: libraries/Standard/src/primitive/resample.birch [hunk 23]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
  *
  * @return A pair, the first element of which gives the ESS, the second
  * element of which gives the logarithm of the sum of weights.
+ *
+ * !!! note
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:285]:
```suggestion
 * @note
```

File: libraries/Standard/src/primitive/resample.birch [hunk 24]
```diff
@@ -259,20 +281,59 @@ function cumulative_weights(w:Real[_]) -> Real[_] {
...
-    let N <- length(w);
-    let W <- 0.0;
-    let W2 <- 0.0;
-    let mx <- nan_max(w);    
-    for n in 1..N {
-      let v <- nan_exp(w[n] - mx);
-      W <- W + v;
-      W2 <- W2 + v*v;
+    // Running maximum of log weights
+    let mx <- -inf;
```
Comment (devmotion) [inline libraries/Standard/src/primitive/resample.birch:300]:
```suggestion
    let mx <- -inf; // running maximum of log weights
```

File: tests/Test/src/basic/test_basic_logsumexp.birch (hunk 1 above)
Comment (devmotion) [reply tests/Test/src/basic/test_basic_logsumexp.birch]: I renamed all occurrences of `logsumexp` (also in the filename) to `log_sum_exp`.

**Comment by lawmurray** [comment]
Thanks for all the revisions @devmotion. Looks good. I've merged back to `numeric` and then `master` now. Thanks again for all the recent contributions!
